### DISCUSSION TITLE
Usage of Handlebars templates

### DISCUSSION THREAD

**Discussion by chkpnt** [webdiscus/html-bundler-webpack-plugin#100] [Q&A] [answered]
Currently, I'm using the following rule:

```javascript
            {
                test: /\.hbs$/,
                use: [
                    {
                        loader: "handlebars-loader",
                        options: {
                            runtime: "handlebars/runtime",
                            inlineRequires: /(\/img\/|\/_svg\/)/,
                        },
                    },
                    {
                        loader: "extract-loader",
                    },
                    {
                        loader: "html-loader",
                        options: {
                            esModule: false,
                            minimize: false,
                            preprocessor: (content, loaderContext) => {
                                var replaceWithSvg = function (arg1, arg2) {
                                    return fs.readFileSync("./src/_svg/" + arg2, "utf8", function (err, data) {
                                        if (err) {
                                            return "";
                                        }

                                        return data.fulltrim();
                                    });
                                };

                                content = content.replace(/<!-- @insertSvg (.*) -->/g, replaceWithSvg);

                                return content;
                            },
                            sources: false,
                        },
                    },
                ],
            },
```

I'm trying to move to `preprocessor: 'handlebars'`, but how would I integrate my custom preprocessor which replaces `<!-- @insertSvg (.*) -->` with the svg's file content?

I've also issues with my helpers: Currently, all of my helpers are in one file `handlebars-helper.js`, where they are registered using `Handlebars.registerHelper(...)`. `handlebars-helpers.js` is imported in my entry point `app.js`. When using `preprocessor: 'handlebars'`, this doesn't work: It seems the helper needs to be available during the compilation.

**Comment by webdiscus** [comment]
Hallo @chkpnt,

Thanks for the question.
I will try to reproduce your interesting use case and find a solution.

**Reply by webdiscus** [reply]
As temporary solution, you can insert SVG-content into HTML using `inline SVG` feature:

Add to Webpack config the rule:
```js
module: {
  rules: [
    {
      test: /\.(png|jpe?g|svg|webp|ico)$/i,
      oneOf: [
        // inline image using `?inline` query
        {
          resourceQuery: /inline/,
          type: 'asset/inline',
        },
        // load image as file
        {
          type: 'asset/module',
          parser: {
            dataUrlCondition: {
              maxSize: 1024,
            },
          },
          generator: {
            filename: 'assets/img/[name].[hash:8][ext]',
          },
        },
      ],
    },
  ],
}
```

In hbs template insert all your SVG source filenames with `?inline` query (exactly as in webpack rule):
```html
<img src="./src/_svg/pic1.svg?inline">
<img src="./src/_svg/pic2.svg?inline">
...
```

The bundler plugin replaces the source img tag with SVG content:
```html
<svg>
  ... pic1 ...
</svg>
<svg>
  ... pic2 ...
</svg>
...
```

**Reply by webdiscus** [reply]
@chkpnt,

You can use your custom helpers, see please [preprocessorOptions.helpers](https://github.com/webdiscus/html-bundler-webpack-plugin?tab=readme-ov-file#loader-option-preprocessor-options-handlebars).

This option can be as array of paths to your helper files or the object containing helper functions.
The helpers must be as a pure function w/o registration. The bundler registrates helpers automatically.

P.S. can you please create a small repo with your use cases and I help you to configure your project.

**Reply by webdiscus** [reply, marked as answer]
@chkpnt

to use your own preprocessor you can use the [beforePreprocessor](https://github.com/webdiscus/html-bundler-webpack-plugin/discussions/100#option-before-preprocessor) callback option or [beforePreprocessor](https://github.com/webdiscus/html-bundler-webpack-plugin/discussions/100#hook-beforePreprocessor) hook.

In the `beforePreprocessor` you can modify the source template content.

See the test case [hook-beforePreprocessor](https://github.com/webdiscus/html-bundler-webpack-plugin/tree/master/test/cases/hook-beforePreprocessor).

**Reply by webdiscus** [reply]
You can see here how works [Hooks & Callbacks](https://github.com/webdiscus/html-bundler-webpack-plugin/discussions/100#plugin-hooks-and-callbacks).

**Reply by chkpnt** [reply]
Thanks for your good input! Using `beforePreprocessor` works fine:

```javascript
            beforePreprocessor: (content, { resourcePath, data }) => {
                if (resourcePath.endsWith(".hbs")) {
                    var replaceWithSvg = function (arg1, arg2) {
                        return fs.readFileSync("./src/_svg/" + arg2, "utf8", function (err, data) {
                            if (err) {
                                return "";
                            }

                            return data.fulltrim();
                        });
                    };

                    content = content.replace(/<!-- @insertSvg (.*) -->/g, replaceWithSvg);

                    return content;
                }
            },
```

But I like the idea to use a `?inline`-query! Sounds to be a cleaner approach.

Regarding the helpers: I can't simply extract them as pure-functions, as my `handlebars-helpers.js` import various other code. But after comparing the bundle bundled with `handlebars-loader` and `html-bundler-webpack-plugin`, I've found the issue: In my `handlebars-helpers.js`, I'm using `import Handlebars from "handlebars/runtime.js";`, but in the preprocessor, you require [`handlebars/dist/handlebars.runtime.min.js`](https://github.com/webdiscus/html-bundler-webpack-plugin/blob/v3.12.0/src/Loader/Preprocessors/Handlebars/index.js#L221). So it seems my helpers were registered on the "wrong" Handlebars instance: After I changed the `import` statement to `handlebars/dist/handlebars.runtime.min.js`, the helpers are working again. :-)

**Reply by chkpnt** [reply]
Maybe the runtime should be configurable as it is in [`handlebars-loader`](https://github.com/pcardune/handlebars-loader/blob/main/index.js#L39).

**Reply by webdiscus** [reply]
> Maybe the runtime should be configurable as it is in [`handlebars-loader`](https://github.com/pcardune/handlebars-loader/blob/main/index.js#L39).

Can you please provide some text for the README: for what this option would be useful, including a simple use case?
Of cause I can add the `runtime` option, but I would like to explain to users why this option is necessary at all.

**Reply by chkpnt** [reply]
When I took over the project I'm working on, neither a package manager nor Webpack was used and all the dependencies resided directly in the repository. In such cases you can't be sure if the dependency, e.g. Handlebars, has been modified. But even if the library hasn't been modified, in my opinion it is better to modernize such a legacy applications step by step. This means: first to introduce a bundler like Webpack, and afterwards to introduce a package manager.

Therefore I was happy that `handlebars-loader` gave me the option to modify the runtime. I was using this configuration at that time:

```javascript
module: {
  // ...
  rules: [
    // ...
    {
      test: /\.hbs$/,
      use: [
        {
          loader: "handlebars-loader",
          options: {
            runtime: "handlebars/runtime", // so that the alias below is used and not the npm module
          },
        },
        {
          loader: "extract-loader",
        },
        {
          loader: "html-loader",
          options: {
            preprocessor: (content, loaderContext) => {
              // ...
            },
          },
        },
      ],
    },
  ],
},
resolve: {
  alias: {
    "@img": path.resolve(__dirname, "src/img"),
    underscore: path.resolve(__dirname, "src/js/_vendors/underscore.min.js"),
    "handlebars/runtime": path.resolve(__dirname, "src/js/_vendors/handlebars.runtime-v4.1.2.js"),
    backbone: path.resolve(__dirname, "src/js/_vendors/backbone.min.js"),
    jquery: path.resolve(__dirname, "src/js/_vendors/jquery-3.4.1.js"),
  },
},
```

But regardless of that, I don't like hard-depending / importing the minified version of a dependency as this makes debugging more sophisticated. Having the option to overwrite the runtime dependency would me allow to use the unminified version of Handlebars. :-)

**Reply by webdiscus** [reply]
@chkpnt 
thank you, the `runtime` option will be added in the next version, at end of this week.

**Reply by webdiscus** [reply]
@chkpnt 

The 3.16.0 version is released.
Now defaults runtime library is `handlebars/runtime`.
Added the [runtime](https://github.com/webdiscus/html-bundler-webpack-plugin?tab=readme-ov-file#loader-option-preprocessor-options-handlebars) option.

P.S.: thank you for the donation!
